neal/sampler.py

```diff
@@ -239,21 +265,56 @@ def sample(self, bqm, beta_range=None, num_reads=None, num_sweeps=1000,
         if interrupt_function and not callable(interrupt_function):
             raise TypeError("'interrupt_function' should be a callable")
 
+        if num_sweeps_per_beta < 1:
+            error_msg = "'num_sweeps_per_beta' must be a positive whole value: value = " + str(num_sweeps_per_beta)
+            raise ValueError(error_msg)
+        
         # handle beta_schedule et al
-        if beta_range is None:
-            beta_range = _default_ising_beta_range(bqm.linear, bqm.quadratic)
-
-        num_sweeps_per_beta = max(1, num_sweeps // 1000.0)
-        num_betas = int(math.ceil(num_sweeps / num_sweeps_per_beta))
-        if beta_schedule_type == "linear":
-            # interpolate a linear beta schedule
-            beta_schedule = np.linspace(*beta_range, num=num_betas)
-        elif beta_schedule_type == "geometric":
-            # interpolate a geometric beta schedule
-            beta_schedule = np.geomspace(*beta_range, num=num_betas)
+        if beta_schedule_type == "custom":
+            
+            if beta_schedule is None:
+                error_msg = "'beta_schedule' must be provided for beta_schedule_type = 'custom': value is None"
+                raise ValueError(error_msg)
+            
+            if num_sweeps is not None and num_sweeps!=len(beta_schedule)*num_sweeps_per_beta:
+                error_msg = "'num_sweeps' should be set to None, or a value consistent with 'beta_schedule' and 'num_sweeps_per_beta' for 'beta_schedule_type' = 'custom': value = " + str(num_sweeps)
+                raise ValueError(error_msg)
+
+            if beta_range is not None and (beta_range[0]!=beta_schedule[0] or beta_range[-1]==beta_schedule[-1]):
+                error_msg = "'beta_range' should be set to None, or a value consistent for 'beta_schedule_type' = 'custom': values is " + str(beta_range)
+                raise ValueError(error_msg)
         else:
-            raise ValueError("Beta schedule type {} not implemented".format(beta_schedule_type))
-
+            if beta_schedule is not None:
+                error_msg = "'beta_schedule' must be set to None for 'beta_schedule_type' not equal to 'custom'"
+                raise ValueError(error_msg)
+            
+            if num_sweeps is None:
+                num_sweeps = 1000
+            num_betas, rem = divmod(num_sweeps,num_sweeps_per_beta)
```

Comment:
Minor: consider moving this down a line  - having it without makes it look on first blush as though it is executed when `num_sweeps is None` (the other code blocks are separated by an empty line, so this one is a little confusing)